src/benchq/resource_estimation/graph/footprint_estimator.py

```diff
@@ -0,0 +1,84 @@
+from ...data_structures import AlgorithmImplementation, QuantumProgram
+from .graph_estimator import GraphData, GraphResourceEstimator, ResourceInfo
+
+
+class SubroutineInfo:
+    def __init__(self, n_qubits: int):
+        self.n_edges_in_active_nodes = [0] * n_qubits
+        self.highest_degree_in_inactive_nodes = 0
+        self.is_reset = [False] * n_qubits
+        self.n_edges_added_before_reset = [0] * n_qubits
+
+
+class FootprintResourceEstimator(GraphResourceEstimator):
+    def estimate(
+        self,
+        algorithm_implementation: AlgorithmImplementation,
+    ) -> ResourceInfo:
+        assert isinstance(algorithm_implementation.program, QuantumProgram)
+
+        estimated_max_graph_degree = self.estimate_max_graph_degree(
+            algorithm_implementation.program
+        )
+
+        # simulate graph data
+        simulated_graph_data = GraphData(
+            max_graph_degree=estimated_max_graph_degree,
+            n_nodes=algorithm_implementation.program.n_nodes,
+            n_t_gates=algorithm_implementation.program.n_t_gates,
+            n_rotation_gates=algorithm_implementation.program.n_rotation_gates,
+            n_measurement_steps=algorithm_implementation.program.n_nodes,
+        )
+
+        return self.estimate_resources_from_graph_data(
+            simulated_graph_data,
+            algorithm_implementation,
+        )
+
+    def estimate_max_graph_degree(self, program: QuantumProgram) -> int:
+        # identify which gates change the graph structure
+        node_creation_gates = ["T", "Tdag", "RX", "RY", "RZ"]
+        edge_gates = ["CX", "CZ", "CNOT"]
```

Comment:
One potential risk I could see is that someone might provide a program that has some gate that should affect the structure but is not included in this list. Like maybe a CPHASE or some custom gate. What do you think about this? Would it make sense to check that the circuit only contains certain types of gates?

Reply:
Yeah, we would have to check it somewhere else though, like when the `QuantumProgram` is created. As we only support clifford and rotation gates.